### ISSUE TITLE
Definition of number of zones 

### ISSUE THREAD

**Issue by diehlpk** [AMReX-Astro/Castro#1663]
Hey, 

I would like to understand the definition of number of zones in Castro, since I am not sure how this value is defined

`Ending run at 15:01:12 UTC on 2021-05-06.
Run time = 127.8339134
Run time without initialization = 126.077113

  Average number of zones advanced per microsecond: 113.552
  Average number of zones advanced per microsecond per rank: 18.925
`
For example, one would run the blast wave with 16777216 cells for 100 hundred time steps and I would assume that the average number of zones processed would be 

16777216 * 100 /  127.8339134 / 1e6 = 13.12422936431828

However, Castro reports a factor of around 10 times more cells. 

Could you please explain the difference, and what do I miss in my calculation above?

**Comment by diehlpk** [comment]
I see that you multiply by the number of the points on the coarse level and I do no understand why this is needed since only the cells in the lowest level are computed or?

https://github.com/AMReX-Astro/Castro/blob/fdeebc501853f578d8c3b4a852d716918bb32004/Source/driver/main.cpp#L174

**Comment by maxpkatz** [comment]
It does include the zones summed over all AMR levels: https://github.com/AMReX-Astro/Castro/blob/fdeebc501853f578d8c3b4a852d716918bb32004/Source/driver/Castro_advance.cpp#L474-L476

We scale by the number of coarse grids when we're accumulating during the run, then undo the scaling at the end when we print the result. There is no specific physical or performance meaning to this scaling, it was just done as an internal trick to avoid the `num_zones_advanced` variable numerically overflowing the size of a double as we accumulate.

**Comment by diehlpk** [comment]
Thanks, my group is interested to reproduce the scaling, you reported for the blast wave on Summit with our astrophysics code (https://github.com/STEllAR-GROUP/octotiger). We are not sure why you use the sum of nodes over all levels and not only the last level to compute the amount of proceed zones?

**Comment by maxpkatz** [comment]
In Castro our AMR strategy involves a coarse grid which represents the entire simulation domain and then fine grids sitting on top of the coarse grid which represent nested subsets of the coarse grid. But we process all zones on every AMR level, even zones that are logically covered by zones on a finer level. This is because our timestep subcycling strategy -- where fine grids advance at smaller timesteps than the coarse grid -- requires having meaningful data for coarse-to-fine interpolation for boundary values on the intermediate fine-grid timesteps. It is true that these covered coarse zones don't represent the final solution, and end up getting overwritten with an average of the fine zones. But that measurement was meant only as a performance metric, and since we do have to computationally evaluate those zones either way, I felt it made sense to include them in the numerator, since they are indirectly included in the denominator. An application that only advanced "leaf" zones would advance "fewer" zones but it would also take less walltime, so to first order this metric could still compare to such an application.

If you have one coarse level and two fine levels, and the refinement ratio between each level is 2, then the calculation for num_zones_advanced in a single full timestep (coarse advance plus advance for all fine levels) is

num_zones_level_0 + 2 * num_zones_level_1 + 4 * num_zones_level_1

By default we take the same number of subcycled timesteps in each level with respect to the level below as the refinement ratio, so that's why in this example a coarse timestep corresponds to 2 level-1 timesteps and 4 level-2 timesteps. Note also that our AMR is dynamic, so the number of zones on level 1 and level 2 can change over time, so over a long simulation if you wanted to calculate it by hand you would have to double check your log to make sure you're accounting for that.

**Comment by diehlpk** [comment]
Ok, thanks that was really helpful to understand how you defined your metric, and we can make a better comparison with Octo-Tiger.

**Comment by maxpkatz** [comment]
Certainly. If you want to do a performance comparison, you might note that our metric does not include time spent in re-gridding the mesh.

**Comment by zingale** [comment]
one additional note: we do the full 12-Riemann corner coupling in the unsplit method, which is more work than the 6-solve method, but the upshot is that it can do bigger timesteps than the 6-solve or a dimensionally-split method.  So depending on what advance you do, a "step' might not mean the same thing.

**Comment by diehlpk** [comment]
Ok, we just used the sedov config file from your examples and edited the file here

`max_step = 100`

`castro.cfl            = 0.4     # cfl number for hyperbolic system`

and we use 

```
# REFINEMENT / REGRIDDING
amr.max_level       = 3      # maximum level number allowed
amr.ref_ratio       = 2 2 2 2 # refinement ratio
amr.regrid_int      = 2       # how often to regrid
amr.blocking_factor = 32       # block factor in grid generation
amr.max_grid_size   = 32
```

Are there any other parameters we should consider?

**Comment by diehlpk** [comment]
Another benchmark we would be interested is a 3d rotating star? 

Is that available in Castro, we could not find it in the examples? 

Only 1D and 2D config files are available there?

**Comment by zingale** [comment]
99% of the time for the rotating star problem is spent in the nuclear physics, so I don't think that is a good problem for comparison, unless you also implement the same reaction network and NSE state.

**Comment by diehlpk** [comment]
Ok, what would be a good problem to compare some astrophysics quantity?

**Comment by zingale** [comment]
For Sedov, why change the CFL number?

**Comment by diehlpk** [comment]
> For Sedov, why change the CFL number?

To have the same on as in our code.

**Comment by zingale** [comment]
I think comparing a fixed number of steps may not be the right approach.  Even at the same CFL number, the algorithms may require different definitions of what the CFL condition is (see the discussion in the Colella 1990 paper, for instance).  So 100 steps might get you to different simulation times even at the same CFL number.  Also, I think we might use an init_shrink in that problem to reduce the timestep a lot initially.

**Comment by shibersag** [comment]
@zingale, isn't it possible to switch nuclear physics off in the rotating star problem?

**Comment by zingale** [comment]
if you just want to test the real EOS + gravity, then I would suggest the StarGrav problem.  It uses the Helmholtz stellar EOS.

**Comment by maxpkatz** [comment]
> Or even better would you mind sharing the config file, you used for the plot in the paper with us?

If you're referring to [the SC20 paper](https://arxiv.org/abs/2007.05218) then the base input deck I used is below. The parameters that were varied were amr.max_grid_size and amr.blocking_factor, and we reported the max and min for each case over a certain range of each. Generally, best performance on GPUs occurs when you have one large box which is as large as can fit on the GPU, for example in this deck we would pair `amr.n_cell=128 128 128` with `amr.blocking_factor=128` and `amr.max_grid_size=128` when running on a single GPU, and the canonical weak scaling test is just increasing n_cell by a factor of 2 in each dimension and also increasing the number of GPUs by a factor of 8, but keeping blocking_factor and max_grid_size fixed. No AMR was used in that test.

```
max_step = 500
stop_time = 1.0e-2

geometry.is_periodic =  0      0      0
geometry.coord_sys   =  0
geometry.prob_lo     =  0      0      0
geometry.prob_hi     =  1.0e9  1.0e9  1.0e9
amr.n_cell           =  128    128    128

castro.lo_bc =  2   2   2
castro.hi_bc =  2   2   2

castro.do_hydro = 1

castro.cfl            = 0.5
castro.change_max     = 1.5

castro.sum_interval   = 1
castro.v              = 1
amr.v                 = 1

amr.max_level       = 0
amr.blocking_factor = 16
amr.max_grid_size   = 32

# Disable plotfiles for benchmarking
amr.checkpoint_files_output = 0
amr.plot_files_output = 0

amr.probin_file = probin
```

**Comment by diehlpk** [comment]
Ok, since the comparison would be more complicated, would you be interested in writing a joint paper comparing the performance / astrophysics results of both codes?

**Comment by maxpkatz** [comment]
We'd be happy to discuss with you in more detail offline (mkatz@nvidia.com) or in this issue to make sure you're running Castro in a way that we think makes sense, and to answer any further questions you have. We could discuss later whether our contributions would rise to the level of co-authorship.
